MB-54131: Geoshape query decode optimization

Likith101 opened the pull request:
- Added buffer pool pointer to polygon
	- Added parent polygon pointer to loop
	- Added decoder function to select, populate and return a pointer to the buffer from a pool of buffers
	- Updated loop decode function to handle chunks of data being read
	- Updated polygon decode function to pass on the buffer pool to the loops
	- Updated extractShapesFromBytes and FilterGeoShapesOnRelation to include the buffer pool in its parameters

Related PRs
        - https://github.com/blevesearch/bleve/pull/1864

Likith101 committed:
MB-54131: Geoshape query decode optimization

	- Added buffer pool pointer to polygon
	- Added parent polygon pointer to loop
	- Added decoder function to select, populate and return a pointer to the buffer from a pool of buffers
	- Updated loop decode function to handle chunks of data being read
	- Updated polygon decode function to pass on the buffer pool to the loops
	- Updated extractShapesFromBytes and FilterGeoShapesOnRelation to include the buffer pool in its parameters

abhinavdangeti commented on geojson/geojson_shapes_util.go:
```diff
@@ -29,11 +29,12 @@ var jsoniter = jsoniterator.ConfigCompatibleWithStandardLibrary
 // the `relation` filter and confirms whether the shape in the document
-//  satisfies the given relation.
+//
+//	satisfies the given relation.
```
Undo this indent, `go fmt ./...` must've changed this because of the unnecessary white space before `satisfies` in line 32 above.

abhinavdangeti commented on geojson/geojson_shapes_util.go:
```diff
@@ -29,11 +29,12 @@ var jsoniter = jsoniterator.ConfigCompatibleWithStandardLibrary
+//	satisfies the given relation.
 func FilterGeoShapesOnRelation(shape index.GeoJSON, targetShapeBytes []byte,
-	relation string, reader **bytes.Reader) (bool, error) {
+	relation string, reader **bytes.Reader, bufPool *[][]byte) (bool, error) {
```
Does this need to a pointer and not simply [][]byte ?

abhinavdangeti commented on s2/loop.go:
```diff
@@ -413,12 +415,12 @@ func (l *Loop) BoundaryEqual(o *Loop) bool {
+//	Then this loop contains XY if and only if Reversed(XY) == the other loop is a hole.
+//	(Intuitively, this checks whether this loop contains a vanishingly small region
+//	extending from the boundary of the other toward the interior of the polygon to
+//	which the other belongs.)
```
Same as above, undo this unrelated indentation change.

abhinavdangeti commented on s2/loop.go:
```diff
@@ -979,21 +981,23 @@ func (l *Loop) ContainsNested(other *Loop) bool {
+//     the loop exterior has been returned, and the integral over the loop
+//     interior can be obtained by adding the integral of f over the entire
+//     unit sphere (a constant) to the result.
 //
```
All the above as well.

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,26 @@ func (l *Loop) decode(d *decoder) {
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+
+	pointsNeeded := int(nvertices) * 3
```
Add a comment here on why `3`.

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1748,10 +1767,12 @@ func (i *intersectsRelation) wedgesCross(a0, ab1, a2, b0, b2 Point) bool {
+//	If ~A contains a point of both B and ~B, then ~A intersects Boundary(B).
+//	So if the intersections of {A, ~A} with {B, ~B} are all non-empty,
+//	the return value is 0, i.e., Boundary(A) intersects Boundary(B).
+//
```
Undo unrelated changes.

abhinavdangeti commented on s2/polygon.go:
```diff
@@ -38,16 +38,16 @@ import (
+//     single loop (see Loop).
 //
-//  - No loop may be empty. The full loop may appear only in the full polygon.
+//   - No loop may be empty. The full loop may appear only in the full polygon.
```
ditto.

abhinavdangeti commented on s2/polygon.go:
```diff
@@ -77,6 +77,8 @@ type Polygon struct {
 	// have a large number of loops, and may be empty for polygons with few loops.
 	cumulativeEdges []int
+
+	BufPool *[][]byte
```
This should perhaps be just `[][]byte`.

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Changed buffer pool from pointer to slice
 - Removed randomly formatted comments

Likith101 committed:
MB-54131: Geoshape query decode

  + add .gitignore

moshaad7 commented on s2/encode.go:
```diff
@@ -222,3 +222,22 @@ func (d *decoder) readUvarint() (x uint64) {
+
+	var buf []byte
+
+	for _, buffer := range buffers {
```
instead of passing around [][]byte as buffers, can we create a new type as GeoBuffPool or something ? 

That way, we can move this logic of selecting a best-fit buffer to some method of that new type.

Thoughts ?

moshaad7 commented on s2/encode.go:
```diff
@@ -222,3 +222,22 @@ func (d *decoder) readUvarint() (x uint64) {
+
+	_, d.err = io.ReadFull(d.r, buf)
+
+	return buf, len(buf) / 8
```
In a pool design, 
you have to make sure that, while a consumer is using an object from the pool, other consumer shouldn't be able to acquire that very same object ( until the first consume return the object back to the pool).

In our case,
If calls to Decode aren't serialised, then we can end-up in situation where multiple decoders are using same buffer.

moshaad7 commented on s2/loop.go:
```diff
@@ -1287,11 +1289,26 @@ func (l *Loop) decode(d *decoder) {
+			break
+		}
+		pointsNeeded = pointsNeeded - pointsRead
+		for j := 0; j < int(pointsRead/3); j++ {
```
We have to make sure that pointsRead is a multiple of 3, 
( I am assuming that since you are creating buffers with sizes as multiple of 3, formula: 8*x*3 , so won't run into a situation where pointsRead isn't multiple of 3 )

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Added a buffer pool struct
 - Increased size of the maximum buffer to 24KB
 - Added better comments

abhinavdangeti commented on s2/buffer_pool.go:
```diff
@@ -0,0 +1,64 @@
+
+func NewGeoBufferPool(maxSize int, minSize int) *GeoBufferPool {
+
+
```
Lose the 2 empty lines here.

abhinavdangeti commented on s2/buffer_pool.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (b *GeoBufferPool) Get(size int) ([]byte) {
+
```
Drop empty line.

abhinavdangeti commented on s2/buffer_pool.go:
```diff
@@ -0,0 +1,64 @@
+
+	for i := range b.buffers {
+		if bufSize <= size {
+			b.buffers[i] = make([]byte, bufSize)
```
It seems we'd be generating a lot of garbage here especially on re-use. Have you considered using a sync.Pool here for the size buffers - so users can return used buffers and we alloc ONLY if the pool has no free buffers of the particular size?

Likith101 replied:
That is an error on my part. Changing this to create the buffer only if it doesn't exist

metonymic-smokey commented on s2/buffer_pool.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+func (b *GeoBufferPool) Get(size int) ([]byte) {
```
I think a buffer pool also requires a corresponding Put() function which will add back buffers into the pool so that they can be re-used later.

moshaad7 replied:
This buffer pool doesn't have a notion of Put.
Because this pool expects Get calls to be serialised.

In future versions we are going to write BufferPool which will be shared across multiple queries, that pool will hold multiple instances of various buffer sizes. 
That pool will be concurrent safe and thus will have the notion of Get and Put.

metonymic-smokey commented on .gitignore:
```diff
@@ -0,0 +1 @@
+.DS_Store
```
Pls remove this file.

moshaad7 replied:
I think there is a need of .gitignore file here because when I pulled the code locally, mac added .DS_Store file as a git change.
Better than manually removing .DS_Store from "git add" , it's better to put it in a ignore file.

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Minor formatting fixes
 - Reusing buffers instead of creating everytime

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
-		l.vertices[i].Z = d.readFloat64()
+
+	// Each vertex requires 24 bytes of storage
+	numBytesNeeded := int(nvertices) * 24
```
Let's declare a `const vertexSize = 24` to use here.

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
+
+		numBytesNeeded = numBytesNeeded - numBytesRead 
+
+		for j := 0; j < int(numBytesRead/24); j++ {
```
Replace 24 with `vertexSize`.

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
+			l.vertices[i+j].Z = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+2) : 8*(j*3+3)]))
+		}
+
+		i = i + int(numBytesRead/24)
```
numBytesRead/vertexSize

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
+		for j := 0; j < int(numBytesRead/24); j++ {
+			l.vertices[i+j].X = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3) : 8*(j*3+1)]))
+			l.vertices[i+j].Y = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+1) : 8*(j*3+2)]))
+			l.vertices[i+j].Z = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+2) : 8*(j*3+3)]))
```
We'll either need to use consts here for 8 and 3 (or derive them from other constants) and more importantly some commentary around why these numbers.

Thejas-bhat commented on s2/buffer_pool.go:
```diff
@@ -0,0 +1,64 @@
+	bufSize := b.maxSize
+
+	for i := range b.buffers {
+		if bufSize <= size {
```
this is my first parse of the PR, but if the caller asks for a buffer size let's say 2047 (Get(2047)), then would you end up returning a buffer of size 1024 instead of 2048?  like isn't the condition for the returned buffer to be something that's closest to the size input but greater than the same? (so that we have allocated enough space and there won't be out of bounds access etc). please let me know if this makes sense

Likith101 replied:
The condition is the largest buffer that is less than or equal to the buffer size. So 2047 should return 1024.

moshaad7 replied:
@Likith101 I kind of agree with what @Thejas-bhat  said.
For a Get(100) I am getting a []byte of size 96. 
I will have to use 2 read system calls to read 100 bytes content from a file ( If I am using a buffer of size 96).
In the buffer pool next bigger item is 96*2=196 bytes item and we should return that to user.

So essentially algorithm should be, 
"smallest buffer with size greater than equal to requested size"

But in case we don't have a buffer which can satisfy the above requirement.
Then in that case, we will return the largest available buffer

Thoughts ?

moshaad7 replied:
But in case we don't have a buffer which can satisfy the above requirement.
Then in that case, we will return the largest available buffer

moshaad7 replied:
```
// returns smallest available buffer of size >= requestd size.
// If no such buffer is available, returns largest buffer from the pool
func (b *GeoBufferPool) Get(size int) []byte {
	currBuffSize := b.minSize
	for i := 0; i < len(b.buffers); i++ {
		if currBuffSize >= size {
			if b.buffers[i] == nil {
				b.buffers[i] = make([]byte, currBuffSize)
			}
			return b.buffers[i]
		}

		currBuffSize = currBuffSize * 2
	}

	// return largest buffer from the pool
	if b.buffers[len(b.buffers)-1] == nil {
		b.buffers[len(b.buffers)-1] = make([]byte, b.maxSize)
	}

	return b.buffers[len(b.buffers)-1]
}
```

abhinavdangeti replied:
Great points by both @Thejas-bhat and @moshaad7 .

I think it makes sense to use a buffer of size **equal to or greater than** what needs to be read up **_until_** the cap (max buffer size in pool) to avoid multiple reads whenever possible.

Likith101 replied:
I have changed the implementation so that the pool returns the smallest buffer that can completely store the number of bytes needed. As a consequence, the decoder will receive a buffer of higher capacity than necessary so I have limited the decoder to read only the required amount of bytes so that it does not read other metadata stored at the end of the file in case of a loop, or other loops in case of a polygon

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Better comments
 - Removed some magic numbers

abhinavdangeti commented on s2/loop.go:
```diff
@@ -1287,11 +1291,36 @@ func (l *Loop) decode(d *decoder) {
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+	floatSize := 8
+	vertexSize := 3 * floatSize
```
declare global constants for floatSize (which definitely needs a better name) and vertexSize.

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Changed buffer pool implementation to give the smallest buffer that fits the data completely
 - Changed decoder implementation to allow buffer sizes bigger than number of bytes needed

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Renamed some constants

moshaad7 commented on s2/encode.go:
```diff
@@ -223,12 +223,16 @@ func (d *decoder) readUvarint() (x uint64) {
+	if size >= len(buf) {
+		_, d.err = io.ReadFull(d.r, buf)
+		return len(buf)
+	} else {
```
since we are using return statement inside "if" block, thus we don't need an else block.

```
if size >= len(buf) {
  _, d.err = io.ReadFull(d.r, buf)
  return len(buf)
}

_, d.err = io.ReadFull(d.r, buf[0:size])
return size

```

Likith101 committed:
MB-54131: Geoshape query decode optimization

 - Added benchmark for loop decode

abhinavdangeti commented on s2/loop.go:
```diff
@@ -24,6 +25,9 @@ import (
 	"github.com/golang/geo/s1"
 )
 
+const SizeOfFloat = 8
```
Should perhaps use reflect to detect size of float64. Will add separately.
